File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.js
class Heap {
    constructor(comparator) {
        this.heap = [];
        this.comparator = comparator; // 최대 힙과 최소 힙은 비교자만 다르고 같은 구조이다.
    }
    
    swap(a, b) {
        [this.heap[a], this.heap[b]] = [this.heap[b], this.heap[a]];
    }
    
    push(item) {
        this.heap.push(item);
        this.bubbleUp();
    }
    
    pop() {
        if (this.size() === 0) return null;
        if (this.size() === 1) return this.heap.pop();
        
        const top = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown();
        
        return top;
    }
    
    bubbleUp() {
        let index = this.heap.length - 1;
        
        while (index > 0) {
            const parent = Math.floor((index - 1) / 2);
            
            // index가 parent보다 우선순위가 낮으면 올리는 것을 멈춘다. 
            if (this.comparator(this.heap[index], this.heap[parent]) >= 0) break; 
            
            this.swap(index, parent);
            index = parent;
        }
    }
    
    bubbleDown() {
        let index = 0;
        // index의 왼쪽 자식이 존재할 동안
        while (index * 2 + 1 < this.heap.length) {
            let child = index * 2 + 1; // 처음 비교 대상은 왼쪽 자식이다.
            
            if (child + 1 < this.heap.length && // 1. 오른쪽 자식이 존재하는가?
                // 2. 오른쪽 자식이 왼쪽 자식보다 우선순위가 높은가? -> 비교 대상을 오른쪽 자식으로 변경
                this.comparator(this.heap[child + 1], this.heap[child]) < 0) child++;
            
            // index가 parent보다 우선순위가 높으면 내리는 것을 멈춘다.
            if (this.comparator(this.heap[index], this.heap[child]) <= 0) break; 
            
            this.swap(index, child);
            index = child;
        }
    }
    
    peek() { return this.heap[0]; }
    
    size() { return this.heap.length; }
}

function solution(operations) {
    
    const minHeap = new Heap((a, b) => a.val - b.val);
    const maxHeap = new Heap((a, b) => b.val - a.val);
    const visited = {}; // 각 숫자의 유효성 체크 (ID 기준)
    
    operations.forEach((op, i) => {
        const [command, value] = op.split(' ');
        const num = Number(value);
        
        if (command === 'I') {
            const node = { val: num, id: i }; // id는 operation의 index이다.
            minHeap.push(node);
            maxHeap.push(node);
            visited[i] = true; // visited가 true여야 유효한 값이다. (삭제되지 않은 값이다.)
        }
        
        if (command === 'D') {
            const targetHeap = num === 1 ? maxHeap : minHeap;
            // 이미 다른 쪽에서 삭제된 노드는 걷어내기 (Lazy Deletion)
            while (targetHeap.size() > 0 && !visited[targetHeap.peek().id]) {
                targetHeap.pop();
            }
            if (targetHeap.size() > 0) {
                const popped = targetHeap.pop();
                visited[popped.id] = false; // 해당 노드를 삭제됨 처리
            }
        }
    })
    
    // 최종 결과 도출 전 유효하지 않은 값 정리
    while (minHeap.size() > 0 && !visited[minHeap.peek().id]) minHeap.pop();
    while (maxHeap.size() > 0 && !visited[maxHeap.peek().id]) maxHeap.pop();

    return minHeap.size() === 0 ? [0, 0] : [maxHeap.peek().val, minHeap.peek().val];
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    min_heap = [] # 최솟값을 위한 힙
    max_heap = [] # 최댓값을 위한 힙 (마이너스 값을 넣어 최소 힙을 최대 힙처럼 사용)
    visited = {}  # ID를 키로 하여 유효성 체크 (True: 유효, False: 삭제됨)

    for i, op in enumerate(operations):
        command, value = op.split()
        num = int(value)

        if command == 'I':
            # 각 노드에 고유 ID(i)를 부여하여 두 힙 사이의 동기화 문제 해결
            heapq.heappush(min_heap, (num, i))
            heapq.heappush(max_heap, (-num, i))
            visited[i] = True
            
        elif command == 'D':
            # num이 1이면 최대 힙에서 삭제, -1이면 최소 힙에서 삭제
            target_heap = max_heap if num == 1 else min_heap
            
            # [지연 삭제 처리] 
            # 힙의 top에 있는 놈이 이미 다른 쪽에서 삭제된 놈이라면 계속 뽑아버림
            while target_heap and not visited[target_heap[0][1]]:
                heapq.heappop(target_heap)
            
            # 유효한 값이 남아있다면 하나 삭제
            if target_heap:
                popped = heapq.heappop(target_heap)
                visited[popped[1]] = False # 삭제된 ID 표시

    # 모든 명령 수행 후, 힙 상단에 남아있는 '이미 삭제된 데이터'들을 최종 정리
    while min_heap and not visited[min_heap[0][1]]:
        heapq.heappop(min_heap)
    while max_heap and not visited[max_heap[0][1]]:
        heapq.heappop(max_heap)

    # 힙이 비어있으면 [0, 0], 값이 있으면 [최댓값, 최솟값] 반환
    if not min_heap:
        return [0, 0]
    else:
        # max_heap은 마이너스를 붙여 넣었으므로 꺼낼 때 다시 -를 붙여 복구
        return [-max_heap[0][0], min_heap[0][0]]